govcd/metadata_v2.go

```diff
@@ -848,23 +869,156 @@ func mergeMetadataAndWait(client *Client, requestUri string, metadata map[string
+// alter the input metadata, but returns a copy of the filtered metadata.
+func filterMetadata(allMetadata *types.Metadata, href, objectName string, metadataToIgnore []IgnoredMetadata) (*types.Metadata, error) {
+	if len(metadataToIgnore) == 0 {
+		return allMetadata, nil
+	}
+
+	result := &types.Metadata{
+		XMLName:       allMetadata.XMLName,
+		Xmlns:         allMetadata.Xmlns,
+		HREF:          allMetadata.HREF,
+		Type:          allMetadata.Type,
+		Xsi:           allMetadata.Xsi,
+		Link:          allMetadata.Link,
+		MetadataEntry: nil,
+	}
+
+	var filteredMetadata []*types.MetadataEntry
+	for _, originalEntry := range allMetadata.MetadataEntry {
+		_, err := filterSingleMetadataEntry(originalEntry.Key, href, objectName, &types.MetadataValue{Domain: originalEntry.Domain, TypedValue: originalEntry.TypedValue}, metadataToIgnore)
+		if err == nil || !strings.Contains(err.Error(), "ignored") {
+			filteredMetadata = append(filteredMetadata, originalEntry)
+		}
+	}
+	result.MetadataEntry = filteredMetadata
+	return result, nil
+}
+
+// filterSingleMetadataEntry filters a single metadata entry given a slice of metadata that needs to be ignored. It doesn't
+// alter the input metadata, but returns a copy of the filtered metadata.
+func filterSingleMetadataEntry(key, href, objectName string, metadataEntry *types.MetadataValue, metadataToIgnore []IgnoredMetadata) (*types.MetadataValue, error) {
+	if len(metadataToIgnore) == 0 {
+		return metadataEntry, nil
+	}
+
+	objectType, err := getMetadataObjectTypeFromHref(href)
+	if err != nil {
+		return nil, err
+	}
+	for _, entryToIgnore := range metadataToIgnore {
+		if entryToIgnore.ObjectType == nil && entryToIgnore.ObjectName == nil && entryToIgnore.KeyRegex == nil && entryToIgnore.ValueRegex == nil {
+			continue
+		}
+		util.Logger.Printf("[DEBUG] Comparing metadata with key '%s' with ignored metadata filter '%s'", key, entryToIgnore)
+		// We apply an optimistic approach here to simplify the conditions, so the metadata entry will always be ignored unless the filters
+		// tell otherwise, that is, if they are nil (not all of them as per the condition above), if they're empty or if they don't match.
+		// All the filtering options (type, name, keyRegex and valueRegex) must compute to true for the metadata to be ignored.
+		if (entryToIgnore.ObjectType == nil || strings.TrimSpace(*entryToIgnore.ObjectType) == "" || *entryToIgnore.ObjectType == objectType) &&
+			(entryToIgnore.ObjectName == nil || strings.TrimSpace(*entryToIgnore.ObjectName) == "" || strings.TrimSpace(objectName) == "" || *entryToIgnore.ObjectName == objectName) &&
+			(entryToIgnore.KeyRegex == nil || entryToIgnore.KeyRegex.MatchString(key)) &&
+			(entryToIgnore.ValueRegex == nil || entryToIgnore.ValueRegex.MatchString(metadataEntry.TypedValue.Value)) {
+			return nil, fmt.Errorf("the entry with key '%s' and value '%v' is being ignored", key, metadataEntry.TypedValue.Value)
```

Comment:
Should there be a single log line to say what this error message passes down the line - should make it easier to debug looking at logs

Reply:
Added. Let me know if the message is informative enough.